apps/src/blockly/addons/cdoFieldNumber.ts

```diff
@@ -0,0 +1,162 @@
+   * @param min Minimum value. Will only be used if config is not
+   *     provided.
+   * @param max Maximum value. Will only be used if config is not
+   *     provided.
+   * @param precision Precision for value. Will only be used if config
+   *     is not provided.
+   * @param validator A function that is called to validate changes to the
+   *     field's value. Takes in a number & returns a validated number, or null
+   *     to abort the change.
+   * @param config A map of options used to configure the field.
+   *     See the [field creation documentation]{@link
+   * https://developers.google.com/blockly/guides/create-custom-blocks/fields/built-in-fields/number#creation}
+   * for a list of properties this parameter supports.
+   */
+  constructor(
+    value?: string | number,
+    min?: string | number | null,
+    max?: string | number | null,
+    precision?: string | number | null,
+    validator?: FieldNumberValidator | null,
+    config?: FieldNumberConfig
+  ) {
+    super(value, min, max, precision, validator, config);
+    this.angleHelper = null;
+  }
+
+  /**
+   * If this field is attached to a block whose output connection is attached to a
+   * connection that has the specified field helper, get the options for that
+   * field helper.
+   * @param {string} fieldHelper - the field helper to retrieve. One of
+   *        Blockly.BlockFieldHelper
+   * @return {Object|undefined} the options object if it exists
+   */
+  getFieldHelperOptions(fieldHelper: string) {
+    return (this.sourceBlock_ &&
+      this.sourceBlock_.outputConnection &&
+      this.sourceBlock_.outputConnection.targetConnection &&
+      (
+        this.sourceBlock_.outputConnection
+          .targetConnection as ExtendedConnection
+      ).getFieldHelperOptions(fieldHelper)) as FieldHelperOptions | undefined;
+  }
+
+  shouldShowAngleHelper() {
+    return this.getFieldHelperOptions(Blockly.BlockFieldHelper.ANGLE_HELPER);
+  }
+
+  protected showEditor_(e?: Event, quietInput?: boolean): void {
+    super.showEditor_(e, quietInput);
+    if (this.shouldShowAngleHelper()) {
+      this.initializeAngleHelper();
+    }
+  }
+
+  /**
+   * Initialize the angle helper.
+   * @private
+   */
+  private initializeAngleHelper() {
+    const sourceBlock = this.getSourceBlock();
+    this.angleHelper = new Blockly.AngleHelper(this.getDirection(), {
+      arcColour: (sourceBlock as ExtendedBlockSvg)?.style.colourPrimary,
+      onUpdate: this.updateAngleValue.bind(this),
+      angle: parseInt(`${this.getValue()}`),
+      enableBackgroundRotation: true,
+    });
+    Blockly.WidgetDiv.createDom();
+    const angleHelperDiv = this.createAngleHelperContainer();
+    Blockly.WidgetDiv.getDiv()?.appendChild(angleHelperDiv);
+    this.angleHelper.init(angleHelperDiv);
+  }
+
+  /**
+   * Get the direction from either the hardcoded setting or the direction field.
+   * @returns {string} The direction value.
+   */
+  getDirection(): string {
```

Comment:
Is this only used when there's an angle helper? If so, it might be useful to rename this or update the comment to explain that.